feat: support kexec from uki

frezbo opened the pull request:
Support kexec from UKI for non-secureboot by extracting kernel, initramfs and cmdline from UKI

Fixes: #10189

frezbo commented:
Currently WIP, doesn't work

smira commented on internal/app/machined/pkg/runtime/v1alpha1/v1alpha1_sequencer_tasks.go:
```diff
@@ -1901,7 +1911,113 @@ func KexecPrepare(_ runtime.Sequence, data any) (runtime.TaskExecutionFunc, stri
+					},
+				},
+				func() error {
+					var kernelFd int
```
I think we need to refactor this a bit, this already becomes a bit of a function that is super hard to read.

we might want to extract these functions out, we can even make them part of the `bootloader` interface

frezbo replied:
:+1: yup, makes sense

smira reviewed (commented):
I guess we need to promote this to uki pipeline to prove that it works?

dsseng commented on internal/app/machined/pkg/runtime/v1alpha1/bootloader/bootloader.go:
```diff
@@ -27,6 +28,9 @@ type Bootloader interface {
 	// RequiredPartitions returns the required partitions for the bootloader.
 	RequiredPartitions() []partition.Options
+
+	// KexecLoad does a kexec using the bootloader.
```
```suggestion
	// KexecLoad does a kexec using the config from bootloader.
```

Maybe another rephrase, but the point is bootloader isn't used, but only its configs

smira replied:
I would say we need to phrase that `KexecLoad` emulates boot as if it was done by the bootloader - we pick the entry to boot based on bootloader information, and then do the boot as if the machine was rebooted

dsseng commented on internal/app/machined/pkg/runtime/v1alpha1/v1alpha1_sequencer_tasks.go:
```diff
@@ -1877,6 +1872,12 @@ func KexecPrepare(_ runtime.Sequence, data any) (runtime.TaskExecutionFunc, stri
 		}
 
+		if efi.GetSecureBoot() {
+			log.Print("kexec skipped as secure boot is enabled")
```
kexec should have secure boot support iirc, yet this can be added later

frezbo replied:
kexec works with secureboot, but that would break disk encryption aka trusted boot, also our UKI doesn't sign kernel if I'm right

frezbo replied:
yeh, it's going to be a mess, since we also extend PCR from init, otherwise we need complex logic to see if we kexeced or not

dsseng commented on internal/app/machined/pkg/runtime/v1alpha1/bootloader/grub/grub.go:
```diff
@@ -44,6 +56,79 @@ func NewConfig() *Config {
+		[]mount.Spec{
+			{
+				PartitionLabel: constants.BootPartitionLabel,
+				FilesystemType: partition.FilesystemTypeXFS,
```
nit: this should perhaps have a constant somewhere, as that's not necessarily XFS but rather we create it as XFS

frezbo replied:
didn't quite get that

smira replied:
I think Dmitry's point is that GRUB has this multiple types around the code (as it does similar mount operations), so the whole block might be shared?

dsseng commented on internal/app/machined/pkg/runtime/v1alpha1/bootloader/grub/grub.go:
```diff
@@ -44,6 +56,79 @@ func NewConfig() *Config {
+
+			fd := int(kernel.Fd())
+
+			// on arm64 we need to extract the kernel from the zboot image if it's compressed
+			if goruntime.GOARCH == "arm64" {
+				var fileCloser io.Closer
+
+				fd, fileCloser, err = zboot.Extract(kernel)
+				if err != nil {
+					return err
+				}
+
+				defer func() {
+					if fileCloser != nil {
+						fileCloser.Close() //nolint:errcheck
+					}
+				}()
+			}
```
This code is shared with sdboot and can be refactored into a new function

frezbo replied:
but it makes more sense to keep it in here only, since it replaces the `fd` and passing it around might be confusing

smira commented:
> I guess we need to promote this to uki pipeline to prove that it works?

I don't see any labels so far...

frezbo commented:
> > I guess we need to promote this to uki pipeline to prove that it works?
> 
> I don't see any labels so far...

promoted the misc-0 and removed when i rebased, though the tests doesn't verify anything since if kexec fails it just continues with a normal reboot, I guess we could store that state in meta or something that it was kexeced

smira reviewed (approved):
once it passes the tests

frezbo commented:
/m
